### PR TITLE
Additional check for leading 0's in number parser

### LINKED ISSUE #2947: Number literals with leading zeros cause issues in JSON marshaling
## Expected Behavior
Run OPA server, specifying the policies 
```
$ opa run -s policies/
```
Then list all policies 
```
$ curl localhost:8181/v1/policies  -H 'Content-Type: application/json'
```
## Steps to Reproduce the Problem
Got  the following compliant
```
$ curl localhost:8181/v1/policies  -H 'Content-Type: application/json'
{
  "code": "internal_error",
  "message": "json: error calling MarshalJSON for type ast.Body: json: error calling MarshalJSON for type *ast.Term: json: error calling MarshalJSON for type ast.Number: json: invalid number literal \"00\""
}
```

## Additional Info

I am wondering whether there are any requirements/dependencies on the policy file or anything else. Thanks.

### PR DESCRIPTION
**Pull request by LCartwright**
Additional checks for leading 0's are now part of the
*Parser.parseNumber()'s validations, prior to parsing.

Fixes #2947

Signed-off-by: Laurence Cartwright <me@lcartwright.com>

### CODE REVIEW

**Review by srenatus — DISMISSED** [review]
We had been discussing (internally, sorry) that this could happen in the scanner instead. But since you've got a test here, and it's doing the trick, I think all that is left is to say **thanks!** and get this merged. Thank you! 😃

**Comment by LCartwright** [comment]
Fantastic, not to worry! If there's a better implementation in the near term on the horizon, I'm more than happy for this little PR to disappear ;) Though I would be happier if it were merged of course!

File: ast/parser.go [hunk 1]
```diff
@@ -867,6 +867,12 @@ func (p *Parser) parseNumber() *Term {
 		}
 	}
 
+	// Check for leading 0's, parsed by math/big.Float.Parse as decimal 0: https://golang.org/pkg/math/big/#Float.Parse
+	if len(prefix) == 0 && len(p.s.lit) > 1 && p.s.lit[0] == '0' && p.s.lit[1] != '.' {
```
Comment (srenatus) [inline ast/parser.go:871]: 💭 What about `-00` (`len(prefix)!=0`) and `0e1`?

Comment (LCartwright) [reply ast/parser.go]: Good point, I've amended the PR and added a test case for these two values

**Comment by tsandall** [comment]
FWIW, I think the change might be a bit simpler if we make it inside the scanner: https://github.com/open-policy-agent/opa/blob/master/ast/internal/scanner/scanner.go#L209. Also, from a correctness point of view, it would be nice if the scanner would not emit invalid tokens -- that just seems wrong.

**Review by srenatus — APPROVED** [review]
Thanks again!

**Pull request merged**